### ISSUE TITLE
`receive(:func).with(time_range)` gives weird error

### ISSUE THREAD

**Issue by eagletmt** [rspec/rspec-mocks#958]
```ruby
RSpec.describe do
  it do
    obj = double('obj')
    t = Time.now
    expect(obj).to receive(:func).with(t .. (t+1))
    obj.func(t .. (t+2))
  end
end
```

```
% bundle exec rspec bug_spec.rb --backtrace
F

Failures:

  1)  should receive func(2015-05-28 14:12:24 +0900..2015-05-28 14:12:25 +0900) 1 time
     Failure/Error: obj.func(t .. (t+2))
     TypeError:
       can't iterate from Time
     # /Users/kohei-suzuki/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/rspec-support-3.2.2/lib/rspec/support/fuzzy_matcher.rb:18:in `each'
     # /Users/kohei-suzuki/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/rspec-support-3.2.2/lib/rspec/support/fuzzy_matcher.rb:18:in `include?'
     # /Users/kohei-suzuki/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/rspec-support-3.2.2/lib/rspec/support/fuzzy_matcher.rb:18:in `include?'
     # /Users/kohei-suzuki/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/rspec-support-3.2.2/lib/rspec/support/fuzzy_matcher.rb:18:in `==='
     # /Users/kohei-suzuki/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/rspec-support-3.2.2/lib/rspec/support/fuzzy_matcher.rb:18:in `values_match?'
     # /Users/kohei-suzuki/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/rspec-support-3.2.2/lib/rspec/support/fuzzy_matcher.rb:31:in `block in arrays_match?'
     # /Users/kohei-suzuki/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/rspec-support-3.2.2/lib/rspec/support/fuzzy_matcher.rb:30:in `each'
     # /Users/kohei-suzuki/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/rspec-support-3.2.2/lib/rspec/support/fuzzy_matcher.rb:30:in `all?'
     # /Users/kohei-suzuki/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/rspec-support-3.2.2/lib/rspec/support/fuzzy_matcher.rb:30:in `arrays_match?'
     # /Users/kohei-suzuki/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/rspec-support-3.2.2/lib/rspec/support/fuzzy_matcher.rb:12:in `values_match?'
     # /Users/kohei-suzuki/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/rspec-mocks-3.2.1/lib/rspec/mocks/argument_list_matcher.rb:58:in `args_match?'
     # /Users/kohei-suzuki/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/rspec-mocks-3.2.1/lib/rspec/mocks/message_expectation.rb:398:in `matches?'
     # /Users/kohei-suzuki/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/rspec-mocks-3.2.1/lib/rspec/mocks/proxy.rb:231:in `block in find_matching_expectation'
     # /Users/kohei-suzuki/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/rspec-mocks-3.2.1/lib/rspec/mocks/proxy.rb:245:in `block in find_best_matching_expectation_for'
     # /Users/kohei-suzuki/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/rspec-mocks-3.2.1/lib/rspec/mocks/proxy.rb:244:in `each'
     # /Users/kohei-suzuki/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/rspec-mocks-3.2.1/lib/rspec/mocks/proxy.rb:244:in `find_best_matching_expectation_for'
     # /Users/kohei-suzuki/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/rspec-mocks-3.2.1/lib/rspec/mocks/proxy.rb:230:in `find_matching_expectation'
     # /Users/kohei-suzuki/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/rspec-mocks-3.2.1/lib/rspec/mocks/proxy.rb:159:in `message_received'
     # /Users/kohei-suzuki/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/rspec-mocks-3.2.1/lib/rspec/mocks/method_double.rb:73:in `proxy_method_invoked'
     # /Users/kohei-suzuki/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/rspec-mocks-3.2.1/lib/rspec/mocks/method_double.rb:60:in `block (2 levels) in define_proxy_method'
     # ./bug_spec.rb:6:in `block (2 levels) in <top (required)>'
     # /Users/kohei-suzuki/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/rspec-core-3.2.3/lib/rspec/core/example.rb:177:in `instance_exec'
     # /Users/kohei-suzuki/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/rspec-core-3.2.3/lib/rspec/core/example.rb:177:in `block in run'
     # /Users/kohei-suzuki/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/rspec-core-3.2.3/lib/rspec/core/example.rb:385:in `block in with_around_and_singleton_context_hooks'
     # /Users/kohei-suzuki/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/rspec-core-3.2.3/lib/rspec/core/example.rb:343:in `block in with_around_example_hooks'
     # /Users/kohei-suzuki/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/rspec-core-3.2.3/lib/rspec/core/hooks.rb:474:in `block in run'
     # /Users/kohei-suzuki/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/rspec-core-3.2.3/lib/rspec/core/hooks.rb:612:in `run_around_example_hooks_for'
     # /Users/kohei-suzuki/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/rspec-core-3.2.3/lib/rspec/core/hooks.rb:474:in `run'
     # /Users/kohei-suzuki/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/rspec-core-3.2.3/lib/rspec/core/example.rb:343:in `with_around_example_hooks'
     # /Users/kohei-suzuki/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/rspec-core-3.2.3/lib/rspec/core/example.rb:385:in `with_around_and_singleton_context_hooks'
     # /Users/kohei-suzuki/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/rspec-core-3.2.3/lib/rspec/core/example.rb:174:in `run'
     # /Users/kohei-suzuki/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/rspec-core-3.2.3/lib/rspec/core/example_group.rb:549:in `block in run_examples'
     # /Users/kohei-suzuki/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/rspec-core-3.2.3/lib/rspec/core/example_group.rb:545:in `map'
     # /Users/kohei-suzuki/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/rspec-core-3.2.3/lib/rspec/core/example_group.rb:545:in `run_examples'
     # /Users/kohei-suzuki/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/rspec-core-3.2.3/lib/rspec/core/example_group.rb:512:in `run'
     # /Users/kohei-suzuki/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/rspec-core-3.2.3/lib/rspec/core/runner.rb:110:in `block (3 levels) in run_specs'
     # /Users/kohei-suzuki/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/rspec-core-3.2.3/lib/rspec/core/runner.rb:110:in `map'
     # /Users/kohei-suzuki/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/rspec-core-3.2.3/lib/rspec/core/runner.rb:110:in `block (2 levels) in run_specs'
     # /Users/kohei-suzuki/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/rspec-core-3.2.3/lib/rspec/core/configuration.rb:1526:in `with_suite_hooks'
     # /Users/kohei-suzuki/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/rspec-core-3.2.3/lib/rspec/core/runner.rb:109:in `block in run_specs'
     # /Users/kohei-suzuki/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/rspec-core-3.2.3/lib/rspec/core/reporter.rb:62:in `report'
     # /Users/kohei-suzuki/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/rspec-core-3.2.3/lib/rspec/core/runner.rb:108:in `run_specs'
     # /Users/kohei-suzuki/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/rspec-core-3.2.3/lib/rspec/core/runner.rb:86:in `run'
     # /Users/kohei-suzuki/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/rspec-core-3.2.3/lib/rspec/core/runner.rb:70:in `run'
     # /Users/kohei-suzuki/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/rspec-core-3.2.3/lib/rspec/core/runner.rb:38:in `invoke'
     # /Users/kohei-suzuki/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/rspec-core-3.2.3/exe/rspec:4:in `<top (required)>'
     # /Users/kohei-suzuki/.rbenv/versions/2.2/bin/rspec:23:in `load'
     # /Users/kohei-suzuki/.rbenv/versions/2.2/bin/rspec:23:in `<main>'
```

**Comment by eagletmt** [comment]
It becomes more problematic when `with` is used multiple times.

```ruby
RSpec.describe do
  it do
    obj = double('obj')
    t = Time.now
    expect(obj).to receive(:func).with(t .. (t+1))
    expect(obj).to receive(:func).with((t+1) .. (t+2))

    obj.func(t .. (t+1))
    obj.func((t+1) .. (t+2))
  end
end
```

It passes with RSpec <= 2.99, but it fails with RSpec >= 3.0.0 saying "can't iterate from Time" .

**Comment by eagletmt** [comment]
I just found a workaround for this issue; adding `eq` matcher.
Is this a recommended way?

```ruby
RSpec.describe do
  it do
    obj = double('obj')
    t = Time.now
    expect(obj).to receive(:func).with(eq(t .. (t+1)))
    expect(obj).to receive(:func).with(eq((t+1) .. (t+2)))

    obj.func(t .. (t+1))
    obj.func((t+1) .. (t+2))
  end
end
```

**Comment by myronmarston** [comment]
This is a duplicate of #938.  It's a bug in ruby.  `time_range === object` works in 1.8.7 but regressed in 1.9+.  I reported it in https://bugs.ruby-lang.org/issues/11113 and the ruby core team promptly fixed it in ruby/ruby@b061634.

For RSpec to work around the bug in Ruby, we'd have to special case time range objects and avoid calling `===` on them...but RSpec uses `===` internally as frequently as it's a generic ruby object matching protocol, so it would be a big pain to specifically avoid this.

> I just found a workaround for this issue; adding eq matcher.

Yep.  On the `eq` matcher, `===` is defined to call `==` on the object so it avoids the problem.  That's actually the exact work around I suggested in #938!

Glad you figured it out :).
